hal/src/common/dmac/dma_controller.rs

```diff
@@ -0,0 +1,293 @@
+    pub fn enable_levels(&mut self, mask: PriorityLevelMask) {
+        // SAFETY This is safe because the use of bitfields ensures that only the
+        // LVLENx bits are written to. The fact that we are given a mask means we need
+        // to do the bit-level setting ourselves.
+        let mask: u16 = mask.into();
+        unsafe {
+            self.dmac.ctrl.modify(|r, w| w.bits(r.bits() | mask));
+        }
+    }
+
+    /// Disable multiple priority levels simultaneously
+    #[inline]
+    pub fn disable_levels(&mut self, mask: PriorityLevelMask) {
+        // SAFETY This is safe because the use of bitfields ensures that only the
+        // LVLENx bits are written to. The fact that we are given a mask means we need
+        // to do the bit-level clearing ourselves.
+        let mask: u16 = mask.into();
+        unsafe {
+            self.dmac.ctrl.modify(|r, w| w.bits(r.bits() & !mask));
+        }
+    }
+
+    /// Enable round-robin arbitration for multiple priority levels
+    /// simultaneously
+    #[inline]
+    pub fn round_robin_arbitration(&mut self, mask: RoundRobinMask) {
+        // SAFETY This is safe because the use of bitfields ensures that only the
+        // RRLVLENx bits are written to. The fact that we are given a mask means we need
+        // to do the bit-level setting ourselves.
+        let mask: u32 = mask.into();
+        unsafe {
+            self.dmac.prictrl0.modify(|r, w| w.bits(r.bits() | mask));
+        }
+    }
+
+    /// Disable round-robin arbitration (ie, enable static priorities) for
+    /// multiple priority levels simultaneously
+    #[inline]
+    pub fn static_arbitration(&mut self, mask: RoundRobinMask) {
+        // SAFETY This is safe because the use of bitfields ensures that only the
+        // RRLVLENx bits are written to. The fact that we are given a mask means we need
+        // to do the bit-level clearing ourselves.
+        let mask: u32 = mask.into();
+        unsafe {
+            self.dmac.prictrl0.modify(|r, w| w.bits(r.bits() & !mask));
+        }
+    }
+
+    /// Release the DMAC and return the register block
+    pub fn free(mut self, _pm: &mut PM) -> DMAC {
+        self.dmac.ctrl.modify(|_, w| w.dmaenable().clear_bit());
+
+        Self::swreset(&mut self.dmac);
+
+        #[cfg(any(feature = "samd11", feature = "samd21"))]
+        {
+            // Disable the DMAC clocking
+            _pm.apbbmask.modify(|_, w| w.dmac_().clear_bit());
+            _pm.ahbmask.modify(|_, w| w.dmac_().clear_bit());
+        }
+
+        // Release the DMAC
+        self.dmac
+    }
+
+    /// Issue a software reset to the DMAC and wait for reset to complete
+    #[inline]
+    fn swreset(dmac: &mut DMAC) {
+        dmac.ctrl.modify(|_, w| w.swrst().set_bit());
+        while dmac.ctrl.read().swrst().bit_is_set() {}
+    }
+
+    /// Split the DMAC into individual channels
+    #[cfg(all(feature = "samd11", not(feature = "max-channels")))]
+    pub fn split(&mut self) -> Channels {
```

Comment:
Won't this conflict with the definition below? It seems like this should just be another `with_num_channels` option added to `mod.rs`.

Reply:
Yeah, this should have been removed when `with_num_channels` was introduced.